hfile_gcs.c

```diff
@@ -37,6 +37,37 @@ DEALINGS IN THE SOFTWARE.  */
 #include "version.h"
 #endif
 
+static const char *
+get_gcs_access_token()
+{
+    // TODO Find the access token in a more standard way
+    char *token = getenv("GCS_OAUTH_TOKEN");
+    if (token) {
+        return token;
+    }
+
+    // Get access token using GOOGLE_APPLICATION_CREDENTIALS env and gcloud auth
+    char access_token[512];
+    token = &access_token[0];
+    int token_size = sizeof access_token;
+    memset(token, 0, token_size);
+    // Check if there is a service account env set
+    const char* service_account = getenv("GOOGLE_APPLICATION_CREDENTIALS");
+    if (service_account && strlen(service_account) > 0) {
+        FILE *fp = popen("gcloud auth application-default print-access-token", "r");
```

Comment:
Two questions: if I understand this correctly, we don't have to worry about the token expiring here because we'll ask for a token before accessing a file every time, correct? 

Second, I tried to find this without success but is there any limit or quota to the token generation? I imagine it would be pretty high, but wonder if it's feasible to run into them (and if we need to worry about caching the token to reuse it...)

Reply:
> I tried to find this without success but is there any limit or quota to the token generation? I imagine it would be pretty high, but wonder if it's feasible to run into them (and if we need to worry about caching the token to reuse it...)
We are setting the environment variable `GCS_OAUTH_TOKEN` for subsequent accesses, cached in a way, so should not hit limits normally. 

> we don't have to worry about the token expiring here because we'll ask for a token before accessing a file every time, correct?
Actually, I have not addressed refreshing the token in this PR, we can alternatively use service accounts via `GOOGLE_APPLICATION_CREDENTIALS` env. The `GCS_OAUTH_TOKEN` could expire just like it could before this change. Let me see if period could be set for a longer duration or if we could refresh the token in some way.

One other thing, this also relies on gcloud installed on the system, I am looking for the REST api, so we can bypass this altogether.

Reply:
There's probably smarter ways around this, but just for the refresh issue I wonder if you could instead set `HTS_AUTH_LOCATION` as suggested here: https://github.com/samtools/htslib/issues/803

Reply:
@mlathara, `HTS_AUTH_LOCATION` supports refresh tokens, but [GCS service accounts do not handle refresh tokens.](https://developers.google.com/identity/protocols/oauth2/service-account). Added code now to explicitly get another token if the previous access token is expired that is 3600s for service accounts.